pkg/reconciler/pipelinerun/resources/pipelinerunstate.go

```diff
@@ -57,10 +58,62 @@ func (state PipelineRunState) IsBeforeFirstTaskRun() bool {
 	return true
 }
 
-// GetNextTasks returns a list of tasks which should be executed next i.e.
+// GetTaskRunsStatus returns a map of taskrun name and the taskrun
+// ignore a nil taskrun in pipelineRunState, otherwise, capture taskrun object from PipelineRun Status
+// update taskrun status based on the pipelineRunState before returning it in the map
+func (state PipelineRunState) GetTaskRunsStatus(pr *v1beta1.PipelineRun) map[string]*v1beta1.PipelineRunTaskRunStatus {
+	status := make(map[string]*v1beta1.PipelineRunTaskRunStatus)
+	for _, rprt := range state {
+		if rprt.TaskRun == nil && rprt.ResolvedConditionChecks == nil {
+			continue
+		}
+
+		var prtrs *v1beta1.PipelineRunTaskRunStatus
+		if rprt.TaskRun != nil {
+			prtrs = pr.Status.TaskRuns[rprt.TaskRun.Name]
+		}
+		if prtrs == nil {
+			prtrs = &v1beta1.PipelineRunTaskRunStatus{
+				PipelineTaskName: rprt.PipelineTask.Name,
+			}
+		}
+
+		if rprt.TaskRun != nil {
+			prtrs.Status = &rprt.TaskRun.Status
+		}
+
+		if len(rprt.ResolvedConditionChecks) > 0 {
+			cStatus := make(map[string]*v1beta1.PipelineRunConditionCheckStatus)
+			for _, c := range rprt.ResolvedConditionChecks {
+				cStatus[c.ConditionCheckName] = &v1beta1.PipelineRunConditionCheckStatus{
+					ConditionName: c.ConditionRegisterName,
+				}
+				if c.ConditionCheck != nil {
+					cStatus[c.ConditionCheckName].Status = c.NewConditionCheckStatus()
+				}
+			}
+			prtrs.ConditionChecks = cStatus
+			if rprt.ResolvedConditionChecks.IsDone() && !rprt.ResolvedConditionChecks.IsSuccess() {
+				if prtrs.Status == nil {
+					prtrs.Status = &v1beta1.TaskRunStatus{}
+				}
+				prtrs.Status.SetCondition(&apis.Condition{
+					Type:    apis.ConditionSucceeded,
+					Status:  corev1.ConditionFalse,
+					Reason:  ReasonConditionCheckFailed,
+					Message: fmt.Sprintf("ConditionChecks failed for Task %s in PipelineRun %s", rprt.TaskRunName, pr.Name),
+				})
+			}
+		}
+		status[rprt.TaskRunName] = prtrs
+	}
+	return status
+}
+
+// getNextTasks returns a list of tasks which should be executed next i.e.
 // a list of tasks from candidateTasks which aren't yet indicated in state to be running and
 // a list of cancelled/failed tasks from candidateTasks which haven't exhausted their retries
-func (state PipelineRunState) GetNextTasks(candidateTasks sets.String) []*ResolvedPipelineRunTask {
+func (state PipelineRunState) getNextTasks(candidateTasks sets.String) []*ResolvedPipelineRunTask {
```

Comment:
should this also be moved to be grouped with the other unexported methods? 🤔

Reply:
Hey @jerop `pipelinerunstate.go` holds attributes to `PipelineRunState` and `PipelineRunFacts`. Have grouped methods based on the receivers:

`PipelineRunState` exported methods
`PipelineRunState` unexported methods
`PipelineRunFacts` exported methods
`PipelineRunFacts` unexported methods

Reply:
sounds great 👍